golang/cmd/mqtt-to-postgresql/database.go

```diff
@@ -87,2346 +92,1383 @@ func PQErrorHandling(sqlStatement string, err error) {
+	zap.S().Debugf("[GetUniqueProductID] assetID: %d, componentName: %s", assetID, componentName)
+	success = false
+	err := statement.SelectIdFromComponentTableByAssetIdAndComponentName.QueryRow(assetID, componentName).Scan(&componentID)
+	if err == sql.ErrNoRows {
+		zap.S().Errorf("No Results Found assetID: %d, componentName: %s", assetID, componentName)
+		return
+	} else if err != nil {
+		PQErrorHandling("GetComponentID() db.QueryRow()", err)
+	}
+	success = true
+	return
+}
 
-		err = item.ToObject(&pt)
+func GetUniqueProductID(aid string, DBassetID uint32) (uid uint32, err error, success bool) {
+	zap.S().Debugf("[GetUniqueProductID] aid: %s, DBassetID: %d", aid, DBassetID)
+	success = false
+	err = statement.SelectUniqueProductIdFromUniqueProductTableByUniqueProductAlternativeIdAndAssetIdOrderedByTimeStampDesc.QueryRow(aid, DBassetID).Scan(&uid)
+	if err == sql.ErrNoRows {
+		zap.S().Errorf("No Results Found aid: %s, DBassetID: %d", aid, DBassetID)
+		return
+	} else if err != nil {
+		PQErrorHandling("GetUniqueProductID db.QueryRow()", err)
+	}
+	success = true
+	return
+}
 
-		if err != nil {
-			err = PQErrorHandlingTransaction("item.ToObject()", err, txn)
-			if err != nil {
-				return
-			}
-		}
-		// Create statement
-		_, err = stmt.Exec(pt.TimestampMs, pt.UID, pt.RecommendationType, pt.Enabled, pt.RecommendationValues, pt.RecommendationTextEN, pt.RecommendationTextDE, pt.DiagnoseTextEN, pt.DiagnoseTextDE)
-		if err != nil {
-			err = PQErrorHandlingTransaction("stmt.Exec()", err, txn)
-			if err != nil {
-				return
-			}
-		}
+func GetLatestParentUniqueProductID(aid string, assetID uint32) (uid int32, success bool) {
+	zap.S().Debugf("[GetLatestParentUniqueProductID] aid: %s, assetID: %d", aid, assetID)
+	success = false
+	err := statement.SelectUniqueProductIdFromUniqueProductTableByUniqueProductAlternativeIdAndNotAssetId.QueryRow(aid, assetID).Scan(&uid)
+	if err == sql.ErrNoRows {
+		zap.S().Errorf("No Results Found aid: %s, assetID: %d", aid, assetID)
+		return
+	} else if err != nil {
+		PQErrorHandling("GetLatestParentUniqueProductID db.QueryRow()", err)
+	}
+	success = true
+	return
+}
+
+// AddAssetIfNotExisting adds an asset to the db if it is not existing yet
+func AddAssetIfNotExisting(assetID string, location string, customerID string) {
 
+	// Get from cache if possible
+	var cacheHit bool
+	_, cacheHit = internal.GetAssetIDFromCache(customerID, location, assetID)
+	if cacheHit { // data found
+		return
 	}
 
-	// Close Statement
-	err = stmt.Close()
+	txn, err := db.Begin()
 	if err != nil {
-		err = PQErrorHandlingTransaction("stmt.Close()", err, txn)
-		if err != nil {
-			return
-		}
+		return
 	}
 
-	// if dry run, print statement and rollback
-	if isDryRun {
-		zap.S().Debugf("PREPARED STATEMENT")
-		err = txn.Rollback()
-		if err != nil {
-			PQErrorHandling("txn.Rollback()", err)
-		}
-		if err != nil {
+	defer func() {
```

Comment:
Can we add a comment before the defer explaining what it does?

`// auto-commit or auto-rollback`